### ISSUE TITLE
env config not working with GCP Batch

### ISSUE THREAD

**Issue by nick-youngblut** [nextflow-io/nextflow#5623]
## Bug report 

My `nextflow.config` contains:

```
env {
    MY_ENV_VAR = "my_value"
}
```

When the executor is `local`, the process using `MY_ENV_VAR` via `os.environ["MY_ENV_VAR"]` runs successfully.
However, when the executor is `google-batch`, the process throws `KeyError: 'MY_ENV_VAR'`. 

https://www.nextflow.io/docs/latest/reference/config.html#env does not state that `env` is not supported for `google-batch`, so I'm assuming that there is either a bug or a lack of docs. 

### Expected behavior and actual behavior

Environmental variables set via the `env` config scope should be available for GCP Batch jobs... or the docs should explicitly state that `env` is not supported for GCP Batch.

### Steps to reproduce the problem

* Set variables in the `env` scope
* Run locally; processes using the env variable should succeed
* Run on GCP Batch; processes using the env variable should fail

### Environment 

* Nextflow version:  24.10.2
* Java version: 21.0.0
* Operating system: Ubuntu 22.04.5
* Bash version: 5.1.16

**Comment by pditommaso** [comment]
I'm unable to replicate this. Can you please provide a test case ?

**Comment by nick-youngblut** [comment]
My reproducible example:

### `main.nf`

```nextflow
workflow { 
    PRINT_ENV()   
    PRINT_ENV_PY()  
}

process PRINT_ENV {
    publishDir file(params.output_dir),  mode: "copy", overwrite: true
    container "ubuntu:latest"
    
    output:
    path "out.txt"

    script:
    """
    echo "$my_env_var" > out.txt
    """
}

process PRINT_ENV_PY {
    publishDir file(params.output_dir),  mode: "copy", overwrite: true
    container "python:3.11"
    
    output:
    path "out_py.txt"

    script:
    """
    test.py > out_py.txt
    """
}
```

### `nextflow.config`

```
params {
    output_dir = "pipeline_output"
}

env {
    my_env_var = "TEST"
}

profiles {
    gcp {
        workDir            = "gs://"###/sandbox/work"
        fusion.enabled     = false
        wave.enabled       = false
        params.output_dir  = "gs://"###/sandbox/pipeline_output"
        process {
            executor       = "google-batch"
            errorStrategy  = "retry"
            maxRetries     = 2
            scratch        = true
        }
        google {
            project   = ""###"
            location  = ""###"
            batch {
                serviceAccountEmail = "###"
                spot                             = true
                maxSpotAttempts     = 3
            }
        }
    }
}
```

### `test.py`

```python
#!/usr/bin/env python3
import os

if __name__ == '__main__':
    print(os.getenv("my_env_var"))
```


### Results

`out.txt` will always contain `TEST` when run locally and on GCP Batch. However, `out_py.txt` will contain `TEST` when run locally and `None` when run on GCP. 


### Notes

If the following is used:

```nextflow
process PRINT_ENV_PY {
    publishDir file(params.output_dir),  mode: "copy", overwrite: true
    container "python:3.11"
    
    output:
    path "out_py.txt"

    script:
    """
    export my_env_var="$my_env_var"
    test.py > out_py.txt
    """
}
```

Then, the env variable value ("TEST") is written out to `out_py.txt`. Importantly, exporting the env variable is only needed when running the pipeline on GCP Batch (versus local).

**Comment by bentsherman** [comment]
Shouldn't the bash version be this?
```groovy
    script:
    """
    echo "\$my_env_var" > out.txt
    """
```

If you don't escape the $ then it will use the global variable `my_env_var`, which can fall back to the environment variables in your config. Though I think we will move away from this pattern in favor of `env('my_env_var')`. Anyway, I think that is why your bash version and modified python version both work.

So the issue seems to be that the config env is not being added to the task environment with Google Batch. And indeed when I check the `.command.run` the config env vars are not there:

```bash
    set -u
    # config env should be here
    [[ $NXF_SCRATCH ]] && cd $NXF_SCRATCH
    export NXF_TASK_WORKDIR="$PWD"
    nxf_stage
```

**Comment by nick-youngblut** [comment]
Yeah, that must be it. I'd have to go back and check on whether I just forgot the escape for `$` in my example, but I'm guessing that I forgot to escape `$` in my code, but found it to still work. I was more hung up on the differences between local and GCP Batch, in regards to the env variable handling.

**Comment by nvnieuwk** [comment]
I've also noticed this issue during one of our runs on GCP. When running the command locally this is in the `.command.run` file:
```
nxf_container_env() {
cat << EOF
export PYTHONNOUSERSITE="1"
export R_PROFILE_USER="/.Rprofile"
export R_ENVIRON_USER="/.Renviron"
export JULIA_DEPOT_PATH="/usr/local/share/julia"
export MOSDEPTH_Q0="NO_COVERAGE"
export MOSDEPTH_Q1="LOW_COVERAGE"
export MOSDEPTH_Q2="CALLABLE"
EOF
}
```

While on GCP it isn't in the `.command.run` file. This is pretty annoying as we needed the `MOSDEPTH_*` variables to be set. This was a pretty deep rabbit hole 😁

**Comment by bentsherman** [comment]
Tracked this down. Fix is in #7609.

The `env` scope never reaches the task on Google Batch because of a method name collision in the plugin.

`BashWrapperBuilder` reads the task environment through `@Delegate private TaskBean bean`, which generates a `getEnvironment()` on the builder, and renders `{{task_env}}` from it:

```groovy
final env = copyStrategy.getEnvScript(environment, runWithContainer)
```

`GoogleBatchScriptLauncher` extends `BashWrapperBuilder` and also implements `GoogleBatchLauncherSpec`, which declares its own unrelated `getEnvironment()`. That one returns the env vars to attach to the Batch `Runnable`, and only the Fusion adapter ever needs them:

```groovy
default Map<String,String> getEnvironment() {
    return Map.of()
}
```

Groovy's `@Delegate` skips generating a method when the class already has an implementation, and an inherited interface default counts as one. So the empty default wins. Reflection confirms it: the declaring class of `getEnvironment()` is `GoogleBatchScriptLauncher`, not `BashWrapperBuilder`.

One shadowed getter explains both halves of what you all saw. The empty map feeds `{{task_env}}`, so no `export` lines land in `.command.run`, and `GoogleBatchTaskHandler` calls the same getter to fill the Batch `Runnable` environment, so it is empty there too. That is exactly the missing block @nvnieuwk spotted.

Fusion runs are unaffected, because `GoogleBatchFusionAdapter` wraps a separate `FusionScriptLauncher` that does not implement the interface, leaving its delegate intact. Everyone reporting here had `fusion.enabled = false`, which fits.

Google Batch is also the only `BashWrapperBuilder` subclass that mixes in an interface. The AWS, Azure, K8s and Fusion launchers extend it and nothing else, which is why this only ever showed up on GCP.

The fix renames the launcher spec method to `getBatchEnvironment()` so the two meanings of "environment" stop colliding. Config env vars then get exported in `.command.run` the same way AWS Batch and Azure Batch already do it, and the Batch `Runnable` environment is unchanged.

Until that ships, the workaround in this thread still holds: reference the config value in the script block so Nextflow interpolates it, e.g. `export MY_ENV_VAR="$MY_ENV_VAR"`.

One note for anyone wondering why this slipped through. Both classes are already `@CompileStatic`, and it compiled clean, because the interface method has a signature identical to the delegate-generated one. Type checking is satisfied, it just binds to the wrong implementation. `@Delegate` has no strict mode and says nothing when it skips a method. I checked that declaring the interface method abstract instead of `default` makes `@Delegate` generate it and the bug disappears, so the `default` body is the whole trigger.

**Comment by nvnieuwk** [comment]
Thanks for the fix! 🥳
